### PR TITLE
Support for HTTP(S) proxy

### PR DESCRIPTION
**Pull request by honzajavorek**
#### :rocket: Why this change?

Dredd should allow and respect HTTP(S) proxy settings:

- Content-Length calculation fixed
- All HTTP requests are now done using the `request` library
- Dredd respects `https_proxy`, `http_proxy`, `no_proxy` environment variables for downloading API description document or for reporting to Apiary
- Proxy **is not** used when testing (this is explained in docs)
- The `createServer` integration tests helper is now able to create also HTTPS server using a self-signed certificate
- Dredd's common options for the `request` library are now at a single place, `dredd.configuration.http`. This makes modifying the settings in testing easier.

#### :memo: Related issues and Pull Requests

- Closes https://github.com/apiaryio/dredd/issues/51
- Closes https://github.com/apiaryio/dredd/issues/751
- Closes https://github.com/apiaryio/dredd/issues/408

#### :white_check_mark: What didn't I forget?

- [x] To write docs
- [x] To write tests
- [x] To put [Conventional Changelog](https://dredd.readthedocs.io/en/latest/contributing/#sem-rel) prefixes in front of all my commits and run `npm run lint`

### CODE REVIEW

File: docs/usage-cli.md [hunk 1]
```diff
@@ -169,6 +169,9 @@ Specifies output file when using additional file-based reporter. This option can
 Additional API description paths or URLs. Can be used multiple times with glob pattern for paths.<br>
 **Default value:** `[]`
 
+### --proxy
+HTTP(S) proxy settings. Overrides http_proxy, HTTP_PROXY, https_proxy, and HTTPS_PROXY environment variables.<br>
```
Comment (kylef) [inline docs/usage-cli.md]: Would it also override `NO_PROXY`?

Comment (honzajavorek) [inline docs/usage-cli.md]: Added to **To Do** to find out. Thanks, good point!

File: src/dredd.coffee [hunk 2]
```diff
@@ -41,6 +42,24 @@ class Dredd
     @transactions = []
     @runner = new Runner(@configuration)
     configureReporters(@configuration, @stats, @tests, @runner)
+    @logProxySettings()
+
+  logProxySettings: ->
+    proxySettings = []
+    for envVariable in Object.keys(process.env)
```
Comment (miiila) [inline src/dredd.coffee]: I'd vote for  `for key, val of process.env`,  but YMMV.

Comment (miiila) [inline src/dredd.coffee]: I'd consider `for for key, val of process.env` more readable and efficient, but YMMV.

File: src/dredd.coffee [hunk 3]
```diff
@@ -41,6 +42,24 @@ class Dredd
...
+      continue unless envVariable.toUpperCase() in PROXY_ENV_VARIABLES
+      continue unless process.env[envVariable] isnt ''
+      proxySettings.push("#{envVariable}=#{process.env[envVariable]}")
+
+    if @configuration.options.proxy
```
Comment (miiila) [inline src/dredd.coffee]: Switch upper if to `if proxySettings.length` would make more sense and could save some rows, but YMMV.

File: src/dredd.coffee [hunk 4]
```diff
@@ -133,12 +152,12 @@ class Dredd
...
   downloadFile: (fileUrl, callback) ->
-    request.get
-      url: fileUrl
-      timeout: 5000
-      json: false
-    , (downloadError, res, body) =>
+    options = {url: fileUrl, timeout: 5000, json: false}
+    options.proxy = @configuration.options.proxy if @configuration.options.proxy
+
+    request.get options, (downloadError, res, body) =>
```
Comment (miiila) [inline src/dredd.coffee]: NIT: Missing parentheses around function call.

File: src/dredd.coffee [hunk 5]
```diff
@@ -133,12 +152,12 @@ class Dredd
     , callback
 
   downloadFile: (fileUrl, callback) ->
-    request.get
-      url: fileUrl
-      timeout: 5000
-      json: false
-    , (downloadError, res, body) =>
+    options = {url: fileUrl, timeout: 5000, json: false}
```
Comment (miiila) [inline src/dredd.coffee]: I somehow don't like the hardcoded timeout, I'd vote for extracting to the constant with some meaningful name at least, but I am not exactly sure about the use case, so it's up to you.

File: src/reporters/apiary-reporter.coffee [hunk 6]
```diff
@@ -177,79 +176,63 @@ class ApiaryReporter
...
-          parsedBody = JSON.parse buffer
-        catch e
-          return callback new Error("Apiary reporter failed to parse Apiary API response body: #{e.message}\n#{buffer}")
-
-        info = {headers: res.headers, statusCode: res.statusCode, body: parsedBody}
-        logger.verbose('Apiary reporter response:', JSON.stringify(info, null, 2))
-        return callback(undefined, res, parsedBody)
-
-    parsedUrl = url.parse @configuration['apiUrl']
+        if CONNECTION_ERRORS.indexOf(err.code) > -1
```
Comment (miiila) [inline src/reporters/apiary-reporter.coffee]: `if err.code in CONNECTION_ERRORS` sounds more coffee-scriptish for me.

File: src/transaction-runner.coffee [hunk 7]
```diff
@@ -450,12 +450,16 @@ class TransactionRunner
...
 
-    return {
+    options =
       uri: url.format(urlObject) + transaction.fullPath
       method: transaction.request.method
       headers: transaction.request.headers
       body: transaction.request.body
-    }
+
+    if @configuration.options.proxy
```
Comment (miiila) [inline src/transaction-runner.coffee]: NIT: you are using `options.proxy = @configuration.options.proxy if @configuration.options.proxy` at https://github.com/apiaryio/dredd/pull/749/commits/02633099e2f070654fcf313335a5347ceec26545#diff-4fe94b7f756f84c61823b246b6738096R156, so you can unify the style, but not big deal.

File: test/integration/helpers.coffee [hunk 8]
```diff
@@ -62,14 +62,19 @@ createServer = ->
 
   app.use(bodyParser.json({size: '5mb'}))
   app.use((req, res, next) ->
+    serverRuntimeInfo.requestedOnce = not serverRuntimeInfo.requested
```
Comment (miiila) [inline test/integration/helpers.coffee]: I don't understand the naming - `once` IMHO tells me, it was done exactly 1 time. But it looks like it's just a negation of `serverRuntimeInfo.requested`.

File: src/dredd.coffee [hunk 9]
```diff
@@ -41,6 +42,24 @@ class Dredd
     @transactions = []
     @runner = new Runner(@configuration)
     configureReporters(@configuration, @stats, @tests, @runner)
+    @logProxySettings()
+
+  logProxySettings: ->
+    proxySettings = []
+    for envVariable in Object.keys(process.env)
+      continue unless envVariable.toUpperCase() in PROXY_ENV_VARIABLES
```
Comment (miiila) [inline src/dredd.coffee]: I saw a `caseless` library for checking the string, so it could be used here as well.

Comment (honzajavorek) [inline src/dredd.coffee]: I could not use `in` in that case. I'd have to iterate over `PROXY_ENV_VARIABLES` and check each item by `caseless.has()`.

**Comment by honzajavorek** [comment]
@miiila Addressed, thanks!

**Pull request merged**